aten/src/ATen/native/Indexing.cpp

```diff
@@ -250,4 +250,50 @@ Tensor & index_put_(Tensor & self, TensorList indices, const Tensor & value) {
   return src.put_(linearIndex, expandedValue);
 }
 
+Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  if (index.dim() != 1) {
+    runtime_error("index_copy_(): Index should have dimension 1 (got %d)", (int) index.dim());
+  }
+  if (dim >= source.dim()) {
+    runtime_error("index_copy_(): Indexing dimension %d is out of bounds of source tensor (%d dimensions)",
+        (int) source.dim());
+  }
+  int64_t numIndices = index.size(0);
+  if (numIndices != source.size(dim)) {
+    runtime_error("index_copy_(): Number of indices (%d) should be equal to source.size(dim) (%d)",
+        (int)numIndices, (int)source.size(dim));
+  }
+  if (index.type().scalarType() != ScalarType::Long) {
+    runtime_error("index_copy_(): Expected LongTensor for index");
+  }
+
+  auto selfSlicedSizes = std::vector<int64_t>(self.sizes());
+  selfSlicedSizes.erase(selfSlicedSizes.begin() + dim);
+  auto sourceSlicedSizes = std::vector<int64_t>(source.sizes());
+  sourceSlicedSizes.erase(sourceSlicedSizes.begin());
+  if (selfSlicedSizes.size() != sourceSlicedSizes.size() ||
+      !std::equal(selfSlicedSizes.begin(), selfSlicedSizes.end(),
+                  sourceSlicedSizes.begin())) {
+    std::stringstream ss;
+    ss << "index_copy_(): Source/destination tensor must have same slice shapes. ";
+    ss << "Destination slice shape: " << selfSlicedSizes << " at dimension " << dim;
+    ss << " and source slice shape: " << sourceSlicedSizes << " at dimension 0.";
+    throw std::runtime_error(ss.str());
+  }
+
+  // Fast path for cuda
+  if (self.is_cuda()) {
+    return self._indexCopy_(dim, index, source);
+  }
+
+  auto workingIndices = index.contiguous();
+  int64_t* index_data = reinterpret_cast<int64_t*>(workingIndices.data_ptr());
+  for (int64_t i = 0; i < numIndices; i++) {
+    auto selfSlice = self.select(dim, index_data[i]);
+    auto sourceSlice = source.select(dim, i);
+    selfSlice.copy_(sourceSlice);
```

Comment:
Quick question: is there a performance regression for 1D arrays? There used to be a dedicated codepath that avoided `select` and `copy`.

Reply:
Possibly. Let me run some numbers.

Reply:
Thanks for catching that, @fmassa, the code is 100x slower now :( :( 

In that case, I'll leave index_copy_ in TH.